skglm/penalties/generalized_linear_cv.py

```diff
@@ -0,0 +1,147 @@
+import numpy as np
+from joblib import Parallel, delayed
+from sklearn.metrics import get_scorer
+from sklearn.utils.extmath import softmax
+from skglm.datafits import Logistic, QuadraticSVC
+from skglm.estimators import GeneralizedLinearEstimator
+
+
+def _kfold_split(n_samples, k, rng):
+    indices = rng.permutation(n_samples)
+    fold_size = n_samples // k
+    extra = n_samples % k
+
+    start = 0
+    for i in range(k):
+        end = start + fold_size + (1 if i < extra else 0)
+        test = indices[start:end]
+        train = np.concatenate([indices[:start], indices[end:]])
+        yield train, test
+        start = end
+
+
+class GeneralizedLinearEstimatorCV(GeneralizedLinearEstimator):
+    """CV wrapper for GeneralizedLinearEstimator."""
+
+    def __init__(self, datafit, penalty, solver, alphas=None, l1_ratio=None,
+                 cv=4, n_jobs=1, random_state=None, scoring=None,
+                 eps=1e-3, n_alphas=100):
+        super().__init__(datafit=datafit, penalty=penalty, solver=solver)
+        self.alphas = alphas
+        self.l1_ratio = l1_ratio
+        self.cv = cv
+        self.n_jobs = n_jobs
+        self.random_state = random_state
+        self.scoring = scoring
+        self.eps = eps
+        self.n_alphas = n_alphas
+
+    def _score(self, y_true, y_pred):
+        """Compute the loss or performance score (lower is better)."""
+        if hasattr(self.datafit, "loss"):
+            return float(np.mean(self.datafit.loss(y_true, y_pred)))
+
+        if self.scoring is not None:
+            if callable(self.scoring):
+                return -float(self.scoring(y_true, y_pred))
+            scorer = get_scorer(self.scoring)
+            return -float(scorer._score_func(y_true, y_pred, **scorer._kwargs))
+
+        if isinstance(self.datafit, (Logistic, QuadraticSVC)):
+            return -float(np.mean(y_true == (y_pred > 0)))
+        return float(np.mean((y_true - y_pred) ** 2))
+
+    def fit(self, X, y):
+        """Fit the model using cross-validation."""
+        if not hasattr(self.penalty, "alpha"):
+            raise ValueError("'penalty' must expose an 'alpha' hyper-parameter.")
+        X, y = np.asarray(X), np.asarray(y)
+        n, p = X.shape
+        rng = np.random.RandomState(self.random_state)
+
+        if self.alphas is not None:
+            alphas = np.sort(self.alphas)[::-1]
+        else:
+            alpha_max = np.max(np.abs(X.T @ y)) / n
+            alphas = np.logspace(
+                np.log10(alpha_max),
+                np.log10(alpha_max * self.eps),
+                self.n_alphas
+            )[::-1]
+        has_l1_ratio = hasattr(self.penalty, "l1_ratio")
+        l1_ratios = [1.] if not has_l1_ratio else np.atleast_1d(
+            self.l1_ratio if self.l1_ratio is not None else [1.])
+
+        folds = list(_kfold_split(n, self.cv, rng))
```

Comment:
you probably can avoid building the full list, that the goal of an enumerator. not crucial here